contracts/Multiwrap.sol

```diff
@@ -0,0 +1,434 @@
+    address private _owner;
+
+    /// @dev The next token ID of the NFT to mint.
+    uint256 public nextTokenIdToMint;
+
+    /// @dev The recipient of who gets the royalty.
+    address private royaltyRecipient;
+
+    /// @dev The percentage of royalty how much royalty in basis points.
+    uint128 private royaltyBps;
+
+    /// @dev Max bps in the thirdweb system
+    uint128 private constant MAX_BPS = 10_000;
+
+    /// @dev Whether transfers on tokens are restricted.
+    bool public isTransferRestricted;
+
+    /// @dev Contract level metadata.
+    string public contractURI;
+
+    /// @dev Token ID => total circulating supply of tokens with that ID.
+    mapping(uint256 => uint256) public totalSupply;
+
+    /// @dev Token ID => total circulating supply of tokens with that ID.
+    mapping(uint256 => uint256) public totalShares;
+
+    /// @dev Mapping from tokenId => uri for tokenId
+    mapping(uint256 => string) private uriForShares;
+
+    /// @dev Mapping from tokenId => wrapped contents of the token.
+    mapping(uint256 => WrappedContents) private wrappedContents;
+
+    /// @dev Checks whether caller has DEFAULT_ADMIN_ROLE.
+    modifier onlyModuleAdmin() {
+        require(hasRole(DEFAULT_ADMIN_ROLE, _msgSender()), "not module admin.");
+        _;
+    }
+
+    /// @dev Initiliazes the contract, like a constructor.
+    function initialize(
+        address _defaultAdmin,
+        string memory _name,
+        string memory _symbol,
+        string memory _contractURI,
+        address _trustedForwarder,
+        address _royaltyRecipient,
+        uint256 _royaltyBps
+    ) external initializer {
+        // Initialize inherited contracts, most base-like -> most derived.
+        __ReentrancyGuard_init();
+        __ERC2771Context_init(_trustedForwarder);
+        __ERC1155_init("");
+
+        // Initialize this contract's state.
+        name = _name;
+        symbol = _symbol;
+        royaltyRecipient = _royaltyRecipient;
+        royaltyBps = uint128(_royaltyBps);
+        contractURI = _contractURI;
+
+        _owner = _defaultAdmin;
+        _setupRole(DEFAULT_ADMIN_ROLE, _defaultAdmin);
+        _setupRole(MINTER_ROLE, _defaultAdmin);
+        _setupRole(TRANSFER_ROLE, _defaultAdmin);
+    }
+
+    ///     =====   Public functions  =====
+
+    /// @dev Returns the module type of the contract.
+    function moduleType() external pure returns (bytes32) {
+        return MODULE_TYPE;
+    }
+
+    /// @dev Returns the version of the contract.
+    function version() external pure returns (uint8) {
+        return uint8(VERSION);
+    }
+
+    /// @dev See ERC1155 - returns the metadata for a token.
+    function uri(uint256 _tokenId) public view override returns (string memory) {
+        return uriForShares[_tokenId];
+    }
+
+    /// @dev Alternative method to get the metadata for a token.
+    function tokenURI(uint256 _tokenId) external view returns (string memory) {
```

Comment:
i think we can phase this out. previously we had issues with opensea, but since then erc1155 has gotten more popular. so we probably don't need it anymore. as uri is part of the spec.